src/controllers/things_controller.js

```diff
@@ -395,74 +400,126 @@ ThingsController.ws('/:thingId/', (websocket, request) => {
   const thingId = request.params.thingId;
   const subscribedEventNames = {};
 
-  function sendMessage(message) {
-    websocket.send(message, (err) => {
+  async function sendMessage(message) {
+    // Convert from internal id to external id
+    if (message.hasOwnProperty('id')) {
+      try {
+        const descr = await Things.getThingDescription(message.id,
+                                                       request.get('Host'),
+                                                       request.secure);
+        message.id = descr.id;
+      } catch (_e) {
+        return;
+      }
+    }
+    websocket.send(JSON.stringify(message), (err) => {
       if (err) {
         console.error(`WebSocket sendMessage failed: ${err}`);
       }
     });
   }
 
   function onPropertyChanged(property) {
-    if (property.device.id !== thingId) {
+    if (typeof thingId !== 'undefined' && property.device.id !== thingId) {
       return;
     }
-    sendMessage(JSON.stringify({
+    sendMessage({
+      id: property.device.id,
       messageType: Constants.PROPERTY_STATUS,
       data: {
         [property.name]: property.value,
       },
-    }));
+    });
   }
 
   function onActionStatus(action) {
-    sendMessage(JSON.stringify({
+    if (action.hasOwnProperty('thingId') &&
+        typeof thingId !== 'undefined' &&
+        action.thingId !== thingId) {
+      return;
+    }
+
+    const message = {
       messageType: Constants.ACTION_STATUS,
       data: {
         [action.name]: action.getDescription(),
       },
-    }));
+    };
+    if (action.hasOwnProperty('thingId')) {
```

Comment:
Is there a reason that actions don't always have a thingId, but events do?

Reply:
The two thingId-less actions are 'pair' which is always global and 'unpair' which includes the thingId as part of its input. It's a bit more complex than it should be, unpair definitely could just include the thingId normally